### PR TITLE
reject pods when under disk pressure

### LINKED ISSUE #84403: why kubelet admits guaranteed pods when under disk pressure?
In the code below, kubelet rejects pods with QOS type guaranteed when under disk pressure, but when kubelet is **also** under memory pressure, those pods are accepted, is this behavior expected? Why is that?

https://github.com/kubernetes/kubernetes/blob/2896fb0be8515e909df71bf29074acf32c54c969/pkg/kubelet/eviction/eviction_manager.go#L128-L164

### PR DESCRIPTION
**Pull request by langyenan**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
kubelet does not reject pods if under both disk pressure and memory pressure

**Which issue(s) this PR fixes**:

Fixes #84403

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

quote from [kubelet-eviction  design proposal](https://github.com/kubernetes/community/blob/ecea14c0af705635a47d6dbc99dc050ed2373065/contributors/design-proposals/node/kubelet-eviction.md#disk-1)
```docs
The kubelet will reject all pods if any of the disk eviction thresholds have been met.
```

### CODE REVIEW

File: pkg/kubelet/eviction/eviction_manager.go [hunk 1]
```diff
@@ -138,7 +138,7 @@ func (m *managerImpl) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAd
 		return lifecycle.PodAdmitResult{Admit: true}
 	}
 	// the node has memory pressure, admit if not best-effort
-	if hasNodeCondition(m.nodeConditions, v1.NodeMemoryPressure) {
+	if hasNodeCondition(m.nodeConditions, v1.NodeMemoryPressure) && len(m.nodeConditions) == 1 {
```
Comment (mattjmcnaughton) [inline pkg/kubelet/eviction/eviction_manager.go]:
```suggestion
        nodeOnlyHasMemoryPressureCondition := hasNodeCondition(m.nodeConditions, v1.NodeMemoryPressure) && len(m.nodeConditions) == 1
	if nodeOnlyHasMemoryPressureCondition {
```

Thoughts on the above small modification? In my mind, it has the advantage of replacing a boolean expression/comment with an explanatory variable name.

**Review by mattjmcnaughton — COMMENTED** [review]
/ok-to-test

Thanks for your pr :)

I believe this fulfills the documented behavior for disk and memory pressure. Would love @gaorong and @tedyu to also take a look, as they commented on the original issue.

Could we please update the unit testing to test this branch?

File: pkg/kubelet/eviction/eviction_manager.go (hunk 1 above)
Comment (langyenan) [reply pkg/kubelet/eviction/eviction_manager.go]: Good suggestion :) , this code style is quite common in k8s code, thanks!

Comment (langyenan) [reply pkg/kubelet/eviction/eviction_manager.go]: Hmm... Never thought a committed suggestion would create a new commit, I'll merge them into one, thank you again :)

**Comment by gaorong** [comment]
lgtm

**Comment by langyenan** [comment]
Working on unit tests

**Review by mattjmcnaughton — COMMENTED** [review]
Conceptually, this looks good to me! I will let someone with slightly more experience on the `pkg/kubelet/eviction/...` portion of the code sign off on the exact unit testing implementation.

Thanks for your work here!

**Comment by dashpole** [comment]
Great find!
/lgtm
/approve

**Pull request merged**